mono/metadata/marshal.c

```diff
@@ -969,63 +923,62 @@ mono_string_builder_to_utf8 (MonoStringBuilder *sb)
  *
  * The return value must be released with mono_marshal_free.
  *
- * This is a JIT icall, it sets the pending exception and returns NULL on error.
+ * This is a JIT icall, it sets the pending exception (in wrapper) and returns NULL on error.
  */
 gunichar2*
-mono_string_builder_to_utf16 (MonoStringBuilder *sb)
+mono_string_builder_to_utf16 (MonoStringBuilderHandle sb, MonoError *error)
 {
-	if (!sb)
+	if (!MONO_HANDLE_BOOL (sb))
 		return NULL;
 
-	ERROR_DECL (error);
-
-	g_assert (sb->chunkChars);
+	g_assert (MONO_HANDLE_GET_BOOL (sb, chunkChars));
 
 	guint len = mono_string_builder_capacity (sb);
 
 	if (len == 0)
-		len = 1; // Why?
+		len = 1; // why?
 
 	gunichar2 *str = (gunichar2 *)mono_marshal_alloc ((len + 1) * sizeof (gunichar2), error);
-	if (!mono_error_ok (error)) {
-		mono_error_set_pending_exception (error);
-		return NULL;
-	}
+	return_val_if_nok (error, NULL);
 
 	str [len] = 0;
-	str [0] = 0; // Cover the case when original len == 0.
+	str [0] = 0; // Cover the case of original len == 0.
 
-	MonoStringBuilder* chunk = sb;
+	MonoArrayHandle chunkChars = MONO_HANDLE_NEW (MonoArray, NULL);
+	MonoStringBuilderHandle chunk = MONO_HANDLE_NEW (MonoStringBuilder, MONO_HANDLE_RAW (sb));
 	do {
-		g_assert (chunk->chunkLength >= 0);
-		if (chunk->chunkLength > 0) {
+		const int chunkLength = MONO_HANDLE_GETVAL (chunk, chunkLength);
+		g_assert (chunkLength >= 0);
+		if (chunkLength > 0) {
 			// Check that we will not overrun our boundaries.
-			gunichar2 *source = (gunichar2 *)chunk->chunkChars->vector;
-
-			g_assert (chunk->chunkOffset >= 0);
-			g_assertf (chunk->chunkLength + chunk->chunkOffset <= len, "A chunk in the StringBuilder had a length longer than expected from the offset.");
-			memcpy (str + chunk->chunkOffset, source, chunk->chunkLength * sizeof (gunichar2));
+			MONO_HANDLE_GET (chunkChars, chunk, chunkChars);
+			const int chunkOffset = MONO_HANDLE_GETVAL (chunk, chunkOffset);
+			g_assert (chunkOffset >= 0);
+			g_assertf ((chunkOffset + chunkLength) >= chunkLength, "integer overflow");
+			g_assertf ((chunkOffset + chunkLength) <= len, "A chunk in the StringBuilder had a length longer than expected from the offset.");
+			gchandle_t gchandle = 0;
+			memcpy (str + chunkOffset, MONO_ARRAY_HANDLE_PIN (chunkChars, gunichar2, 0, &gchandle), chunkLength * sizeof (gunichar2));
```

Comment:
this (and a couple others, but especially this one, since it loops) could be `NO_SAFEPOINTS`